core/rawdb/bor_receipts.go

```diff
@@ -0,0 +1,211 @@
+func ReadRawBorReceipt(db kv.Tx, hash common.Hash, number uint64) *types.Receipt {
+	// Retrieve the flattened receipt slice
+	data := ReadBorReceiptRLP(db, hash, number)
+	if data == nil || len(data) == 0 {
+		return nil
+	}
+
+	// Convert the receipts from their storage form to their internal representation
+	var storageReceipt types.ReceiptForStorage
+	if err := rlp.DecodeBytes(data, &storageReceipt); err != nil {
+		log.Error("Invalid receipt array RLP", "hash", hash, "err", err)
+		return nil
+	}
+
+	return (*types.Receipt)(&storageReceipt)
+}
+
+// ReadBorReceipt retrieves all the bor block receipts belonging to a block, including
+// its correspoinding metadata fields. If it is unable to populate these metadata
+// fields then nil is returned.
+func ReadBorReceipt(db kv.Tx, hash common.Hash, number uint64) *types.Receipt {
+	// We're deriving many fields from the block body, retrieve beside the receipt
+	borReceipt := ReadRawBorReceipt(db, hash, number)
+	if borReceipt == nil {
+		return nil
+	}
+
+	// We're deriving many fields from the block body, retrieve beside the receipt
+	receipts := ReadRawReceipts(db, number)
+	if receipts == nil {
+		receipts = make(types.Receipts, 0)
+	}
+
+	body, _, _ := ReadBody(db, hash, number)
+	if body == nil {
+		log.Error("Missing body but have bor receipt", "hash", hash, "number", number)
+		return nil
+	}
+
+	if err := types.DeriveFieldsForBorReceipt(borReceipt, hash, number, receipts); err != nil {
+		log.Error("Failed to derive bor receipt fields", "hash", hash, "number", number, "err", err)
+		return nil
+	}
+	return borReceipt
+}
+
+// WriteBorReceipt stores all the bor receipt belonging to a block.
+func WriteBorReceipt(tx kv.RwTx, hash common.Hash, number uint64, borReceipt *types.ReceiptForStorage) error {
+	// Convert the bor receipt into their storage form and serialize them
+	bytes, err := rlp.EncodeToBytes(borReceipt)
+	if err != nil {
+		return err
+	}
+
+	// Store the flattened receipt slice
+	if err := tx.Append(kv.BorReceipts, borReceiptKey(number), bytes); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// DeleteBorReceipt removes receipt data associated with a block hash.
+func DeleteBorReceipt(tx kv.RwTx, hash common.Hash, number uint64) {
+	key := borReceiptKey(number)
+
+	if err := tx.Delete(kv.BorReceipts, key, nil); err != nil {
+		log.Crit("Failed to delete bor receipt", "err", err)
+	}
+}
+
+// ReadBorTransactionWithBlockHash retrieves a specific bor (fake) transaction by tx hash and block hash, along with
+// its added positional metadata.
+func ReadBorTransactionWithBlockHash(db kv.Tx, txHash common.Hash, blockHash common.Hash) (*types.Transaction, common.Hash, uint64, uint64, error) {
+	blockNumber, err := ReadBorTxLookupEntry(db, txHash)
+
+	if err != nil {
+		return nil, common.Hash{}, 0, 0, err
+	}
+
+	if blockNumber == nil {
+		return nil, common.Hash{}, 0, 0, nil
+	}
+
+	body, _, _ := ReadBody(db, blockHash, *blockNumber)
```

Comment:
Too expensive to retrieve the entire body just to see the number of transactions `len(body.Transactions)`. There is a cheaper way of doing it, just reading 1 record from `BlockBody` table, which contains the number of transactions